.github/workflows/create-release-experimental.yml

```diff
@@ -0,0 +1,75 @@
+      - main
+    paths:
+      - 'v2/**'
+  # no content, allows manual triggering
+  workflow_dispatch:
+
+env:
+  ARTIFACT_VERSION: experimental
+
+jobs:
+  build-and-push:
+    runs-on: ubuntu-latest
+    steps:
+      - name: Checkout code
+        uses: actions/checkout@v3
+        with:
+          fetch-depth: 0 # required to access tags
+          submodules: "true"
+
+      - name: Build & run devcontainer image
+        # this always builds a new image from scratch rather than from the build-devcontainer-image workflow output
+        # so that we pick up the latest versions of everything
+        # NB: if you update this also update live-validation.yml, pre-release-tests.yaml and create-release-official.yaml
+        id: devcontainer
+        run: |
+          docker build --tag devcontainer:latest .devcontainer
+          mkdir -p $HOME/.docker # in case it doesn't exist
+          container_id=$(docker create -w /workspace -v $GITHUB_WORKSPACE:/workspace -v /var/run/docker.sock:/var/run/docker.sock devcontainer:latest)
+          docker start "$container_id"
+          echo "container_id=$container_id" >> $GITHUB_ENV
+
+      - name: Build required release files
+        run: |
+          container_id=${{env.container_id}}
+          docker exec "$container_id" task VERSION=${{ env.ARTIFACT_VERSION }} LATEST_VERSION_TAG=${{ env.ARTIFACT_VERSION }} make-release-artifacts
+
+      - name: Upload release assets
+        uses: svenstaro/upload-release-action@7319e4733ec7a184d739a6f412c40ffc339b69c7 # this is v2.5.0, but pinned
+        with:
+          repo_token: ${{ secrets.GITHUB_TOKEN }}
+          tag: refs/tags/${{ env.ARTIFACT_VERSION }}
+          file: "v2/out/release/*"
+          file_glob: true
+          prerelease: true
+          overwrite: true
+          release_name: "Experimental"
+          body: |
+            This is an experimental release which containing the most recent commits from the main branch as of commit: ${{ github.sha }}.
+
+            **This release might not be stable. Use at your own risk.** 
+
+            It is only intended for developers wishing to try out the latest feature, some of which may not be fully implemented.
+
+      - name: Docker login
+        run: |
+          container_id=${{env.container_id}}
+          docker exec -e AZURE_CLIENT_ID -e AZURE_CLIENT_SECRET -e DOCKER_REGISTRY "$container_id" task VERSION=${{ env.ARTIFACT_VERSION }} LATEST_VERSION_TAG=${{ env.ARTIFACT_VERSION }} docker-login
+        env:
+          DOCKER_REGISTRY: ${{ secrets.REGISTRY_LOGIN }}
+          AZURE_CLIENT_ID: ${{ secrets.AZURE_CLIENT_ID }}
+          AZURE_CLIENT_SECRET: ${{ secrets.AZURE_CLIENT_SECRET }}
+
+      - name: Build, tag and push docker image
+        run: |
+          container_id=${{env.container_id}}
+          docker exec -e DOCKER_PUSH_TARGET "$container_id" task VERSION=${{ env.ARTIFACT_VERSION }} LATEST_VERSION_TAG=${{ env.ARTIFACT_VERSION }} controller:docker-push-multiarch
```

Comment:
re-tagging different docker images over and over is possible to cause problems, I think. See [the default image pull policy](https://kubernetes.io/docs/concepts/containers/images/#imagepullpolicy-defaulting)

We currently set `imagePullPolicy` to `Always` (which TBH is probably wrong), but would mitigate this issue at least somewhat... though even then there are possible awkward situations with for example the following situations:
1. User installs experimental ASO via YAML.
2. We merge a new PR containing a new resource, including webhooks, Kubernetes RBAC roles, etc
3. Users pod restarts for whatever reason.
4. Users pod pulls the new image, _but_ the resource YAMLs (including RBAC roles, webhooks) are from the N-1st version of experimental. Pod will try to install the new CRD that is now embedded inside of it from the new resource (assuming configuration is `crdPattern=*`), but the pod will crash during launch because it will attempt to start a watch on the resource but it doesn't have permission to do so.

The workaround there is that you can delete the ASO namespace and re-install from latest and that should fix it. The number of times this happens will likely not be _super_ common... but we should probably document this behavior and _really really_ emphasize that people need to not run this in production. It literally won't work if they do it long enough because we will add resources and this will happen and break them.

It's much less of a problem for doing local sorta testing, kicking the tires, etc - but running for a week in a dev or test cluster is more likely to run into this issue.

Possible solutions are to pin the specific image hash...

Reply:
I see your point but it's more of a use-case scenario, tagging with the commit is a good addition to both scenarios which users can use depending on their needs.

Would that make sense? If so, I'll open a follow-up PR with docs + additional image pushed with the sha tag

Reply:
personally I think it would be good to have one with a commit, even if it's just built nightly so that not all commits are built.

If I need a change sooner than the current 2 month release schedule I would like to use a fixed image so I can get changes sooner and provide feedback (and continue running it until the release).

A moving image isn't very useful except for a quick test.

Reply:
These are the two scenarios that I often see in the various projects that I maintain/interact with:
1. `:main`/`:experimental` - I want  to run the latest and greatest, don't ask me to check what the latest commit was
2. `:<sha>` - I want to test a given change and stick to it, if I want another change, then I will look it up

I believe ASO needs both and 2) is what Tim is looking for as well

Reply:
Do you actually need to push a specific SHA-tagged image? I don't think you do because you can always pull a sha-tagged image regardless.

I just tested this locally and it already works:
`docker pull mcr.microsoft.com/k8s/azureserviceoperator@sha256:d82f04240b3b72881342ae1cd0714e154cbaa00bc9f8fb226e7ec8c23fefe28f` - which is the sha for v2.3.0.

Reply:
The only thing is that is needed is that, the YAML in the `experimental` release will be referring to the YAML tag, so users will need to use `kustomize` or some other mechanism to update that to refer to a specific SHA... unless we modify the release to produce a YAML that refs a sha rather than the `experimental` tag directly, which possibly is worth doing but may need some build script changes.

Reply:
Fair enough - We should be good to go then!

With regards to the YAML; if they want to use a SHA instead of experimental tag, then I'd suggest we leave it up to customers to do a find/replace after downloading the artifacts, I would not offer this OOTB personally.